chore: api error handling

saminakh opened the pull request:
This PR adds an additional error modal specifically for handling API errors. For known errors, a friendly message is displayed; otherwise, a detailed modal with the error message, trace id, and contact email is displayed.

Examples:
<img width="1201" alt="Screen Shot 2022-01-07 at 3 17 22 PM" src="https://user-images.githubusercontent.com/8646176/148615250-50036483-75e7-4ef8-9bb6-a6b1b5450595.png">
<img width="1195" alt="Screen Shot 2022-01-07 at 3 18 33 PM" src="https://user-images.githubusercontent.com/8646176/148615195-bb277c4e-083a-4e3f-b49b-626263f50de3.png">
<img width="1197" alt="Screen Shot 2022-01-06 at 11 40 04 PM" src="https://user-images.githubusercontent.com/8646176/148502910-0807a96d-b7ba-45aa-a79e-684acafe9a3b.png">

alexandreaco commented:
I can't leave a review because it's still a WIP branch, but here's some feedback :) @saminakh 

1. https://github.com/radixdlt/olympia-wallet/pull/410/files#diff-7a7a37b12ee1265d7e27577ec286120d2cbc0940908635e264a2be44ccb9a8a0R6-R13 I don't think we'll ever show the transaction building error because isApiError will return true first
2. https://github.com/radixdlt/olympia-wallet/pull/410/files#diff-7deb4a2ae75ee9c0d8e9807092fd413c257fae59ca15a80bb16276daf099577eL244-R244 Typo, we lost a space between words
3. I wonder if we want the copy to say "Unknown Api Error" or something more confident like "Received Api Error" when we don't have friendly text to show. <img width="1312" alt="Screen Shot 2022-01-05 at 9 13 00 AM" src="https://user-images.githubusercontent.com/4480888/148231560-4664c632-768d-411f-8faa-2f758cc27f28.png"> 
4. I'm having trouble creating a "not enough XRD to cover fee" error. This may be an issue with the Api <img width="1440" alt="Screen Shot 2022-01-05 at 9 14 53 AM" src="https://user-images.githubusercontent.com/4480888/148231863-cf41db32-e831-47a5-96a7-9c8c5033998d.png"> 
5. For `NotEnoughTokensForUnstakeError`, we should check with Radix to see if we need to present both of those messages to the user

saminakh commented:
Thank you for the feedback @alexandreaco !
1. Todo
2. Fixed!
3. I think that's a good idea! I changed it!
4. I did manage to reproduce this, it appears that the API only sends this error if you're sending NON-native tokens and you don't have enough native tokens to cover the transaction fee. There doesn't appear to be a corresponding error for if you're sending native tokens though and don't have enough to cover the transaction fee. Should we tell radix and recommend they add an api error for that?
5. Todo

alexandreaco commented on src/components/ErrorModalApi.vue:
```diff
@@ -0,0 +1,120 @@
+    const showDetails = !isApiError(type)
+    let errorMsg: string
+    if (showDetails) {
+      errorMsg = 'Unexpected Error'
```
Let's make sure we're using the i18n text file for these bits of text :)

saminakh replied:
👍

alexandreaco commented on src/components/ErrorModalApi.vue:
```diff
@@ -0,0 +1,120 @@
+    } else if (type === 'BelowMinimumStakeError') {
+      const minimum = error.value.details?.minimum_amount?.value
+      const displayMinimum = minimum ? asBigNumber(minimum) : '10'
+      errorMsg = t(`apiErrors.${type}`, { minimum: displayMinimum })
```
nit: I don't think you need to template the string `apiErrors.${type}` because we always know the type is `BelowMinimumStakeError`. Nice use of custom values in the i18n call!

saminakh replied:
Yeah good point! I did this in another spot too so I'll change 'em both!

alexandreaco commented on src/components/ErrorModalApi.vue:
```diff
@@ -0,0 +1,120 @@
+        errorMsg = t(`apiErrors.${type}.pending`, { pending: asBigNumber(pending) })
+      }
+    } else {
+      errorMsg = t(`apiErrors.${type}`)
```
Would we really ever get to this clause? Maybe you can reverse your conditionals and just leave a single "else"

if (requested <= add(staked, pending))
  // show pending error
else 
  // show default error

saminakh replied:
Oh! I think i named the error messages confusingly. I'll change this to make it clearer!

alexandreaco commented on src/composables/useTransactions.ts:
```diff
@@ -269,8 +273,12 @@ export default function useTransactions (radix: ReturnType<typeof Radix.create>,
-    transactionSubs.add(completion.subscribe(handleTransactionCompleted))
-    transactionSubs.add(events.subscribe(handleTransactionLifecycleEvent))
+    transactionSubs.add(
+      completion.subscribe(handleTransactionCompleted, errorHandler)
```
neat! i didn't know you could pass multiple functions into subscribe

saminakh replied:
I didn't either! someone from radix had a code example in the channel 😄

alexandreaco commented on src/composables/useWallet.ts:
```diff
@@ -85,7 +85,7 @@ const createWallet = (mnemonic: MnemomicT, pass: string, network: Network) => {
     setWallet(newWallet)
     saveDerivedAccountsIndex(0, network)
-    persistNodeUrl('https://sandpitnet-gateway.radixdlt.com')
+    persistNodeUrl('https://stokenet-gateway.radixdlt.com')
```
I think you'll have rebase conflicts because I made these strings a const. sorry!

saminakh replied:
No worries! Rebased 👍

alexandreaco commented on src/text/index.ts:
```diff
@@ -242,6 +242,22 @@ const messages = {
       unableToPrepareUnstakingTransactionPointThree: 'Insufficient XRD to pay the required transaction fee'
+    },
+    apiErrors: {
+      NetworkNotSupportedError: 'The network selected is not valid.',
```
We can remove some of these errors. We should only show custom errors for the following codes

InvalidAccountAddressError
InvalidValidatorAddressError
NotEnoughNativeTokensForFeesError
NotEnoughTokensForTransferError
NotEnoughTokensForStakeError
NotEnoughTokensForUnstakeError
BelowMinimumStakeError
CannotStakeError
MessageTooLongError

saminakh replied:
👍 totally forgot to remove these cases, thanks for the reminder!

alexandreaco commented on src/composables/useErrors.ts:
```diff
@@ -1,20 +1,27 @@
+  'NotEnoughTokensForTransferError',
+  'NotEnoughTokensForStakeError',
+  'NotEnoughTokensForUnstakeError',
+  'BelowMinimumStakeError',
```
We also need to remove the inline check for 90xrd

saminakh replied:
I should have asked about the inline validations, that was silly of me not to. I thought maybe these modals would just be backups if the inline checks failed? But that was a silly assumption in retrospect 😬

alexandreaco commented on src/composables/useErrors.ts:
```diff
@@ -1,20 +1,27 @@
+  'NotEnoughTokensForUnstakeError',
+  'BelowMinimumStakeError',
+  'CannotStakeError',
+  'MessageTooLongError'
```
We also need to remove the inline check for when messages are longer than 160 characters

saminakh replied:
👍

alexandreaco reviewed (changes requested):
Thanks for all of this work! 🙇‍♀️  I left a few notes about small nits and minor updates

saminakh commented:
@alexandreaco for title/copy change for the error modals, I went with:
- Cannot Create Transaction
- Cannot Complete Stake
- Cannot Complete Unstake
Do those sound reasonable?

alexandreaco commented on src/App.vue:
```diff
@@ -3,8 +3,13 @@
     <div class="relative z-10 min-h-screen">
       <template v-if="latestError">
+        <ErrorModalApi
+          v-if="isApiError"
```
`isApiError` returns true if `latest.error?.details?.type` exists. This means that we'll never get to the subsequent else-if statements that are checking for that type to be a specific string. Maybe you intended to import `isApiError` from `useErrors`?

saminakh replied:
Ahh I see what's confusing here. The flow should be: check that `latest.error.details.type` exists. If yes: show api error modal, where we look at the type (and if the type is found, show that simplified error modal for that type, otherwise, show the detailed error modal w/ the trace id). If `latest.error.details.type` doesn't exist, _then_ we think it's not an API error and show one of the other modals instead.

So I think the problem here is, is there a better way to check if it's an API error or not? I can look a little more closely at the error and see if there's a nicer way to do this.

saminakh replied:
Ok! Here's what I did. I:
- updated where we handle the api errors to include a `type` flag of 'api' before adding it to the list of errors for the app. This 1) is probably a better way of checking if the error is an api type or a general app error, and 2) makes the logic a little more clear
- renamed the `isApiError` --> `isKnownApiError`, to make the purpose of that function more clear

alexandreaco replied:
Thanks, this makes much more sense now!

alexandreaco commented on src/components/ErrorModalApi.vue:
```diff
@@ -0,0 +1,135 @@
+        <div class="bg-rRed rounded-full inline-block px-2 py-0.5 m-2 text-white text-sm">{{ errorsCount }}</div>
+      </div>
+      <div v-if="showDetails">
+        <p class="mb-2">Please contact support at <a class="text-rBlue"
```
Can we add these phrases to the text.ts file?

saminakh replied:
Yeah! Should i just break the string into two chunks surrounding the link?

saminakh replied:
(I ended up just doing that!)

alexandreaco commented on src/text/index.ts:
```diff
@@ -242,6 +242,51 @@ const messages = {
+    apiErrors: {
+      unknown: 'Unexpected Error',
+      InvalidAccountAddressError: {
+        message: 'The recipient Radix address specified is not valid.',
```
👍 This nesting is really easy to follow and documents the flow well. Great idea

saminakh replied:
😄

alexandreaco commented on src/views/Wallet/WalletTransaction.vue:
```diff
@@ -96,9 +96,6 @@
                       class="w-full"
                       label="Message"
                       :placeholder="$t('transaction.messagePlaceholder')"
-                      :rules="{
```
I'm seeing a runtime error now that rules is undefined. Can you set rules to an empty object instead?

saminakh replied:
Good catch! Sorry about that, I'll go and fix it!

alexandreaco reviewed (changes requested):
The new error modals look great. Thanks for adding in the titles :) 

I left a few notes in the PR that we should cleanup before merging. 

Lastly, can you add some horizontal padding to this sentence so we don't have an orphaned word on the second line? In the screenshot I tried px-12 and it looks nicer.
<img width="1268" alt="Screen Shot 2022-01-10 at 6 29 58 AM" src="https://user-images.githubusercontent.com/4480888/148759882-f9093429-3285-44bc-8518-8f5323da8bea.png">

alexandreaco reviewed (approved):
💯
